### PR TITLE
[SYCL] Refactor SYCL Post Link Library

### PR DESCRIPTION
**Pull request by maksimsab**
Post Link compilation flow consists of Module splitting, specialization constants processing and ESIMD processing. To correspond to the all functionalities this patch adds renames in order to be less confusing.

The following changes are added:
* ModuleSplitProcessingSettings structure is renamed to ModuleProcessingSettings.
* Added function convertProcessingSettingsToString to add additional piece of information in the testing.
* SplitModule structure is renamed to ProcessedModule.
* runSYCLSplitLibrary is renamed to runSYCLPostLinkLibrary.
* DryRun mode in runSYCLPostLinkLibrary is combined with Verbose mode in order to remove code duplication.
* splitSYCLModule is renamed to SYCLPostLinkProcess.

### CODE REVIEW

**Comment by maksimsab** [comment]
My patch https://github.com/intel/llvm/pull/15133 has become so big locally that I decided to split it in several PRs in order to decrease a burden on reviewers.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp [hunk 1]
```diff
@@ -743,76 +743,76 @@ runSYCLPostLinkTool(ArrayRef<StringRef> InputFiles, const ArgList &Args) {
...
         *ImageFileOrErr, util::PropertySetRegistry(), "")};
     return Modules;
   }
 
-  return llvm::module_split::parseSplitModulesFromFile(*TempFileOrErr);
+  return llvm::module_split::parseProcessedModulesFromFile(*TempFileOrErr);
 }
 
-/// Invokes SYCL Split library for SYCL offloading.
+/// Invokes SYCL Post Link library for SYCL offloading.
```
Comment (bader) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:754]: I don't think "Post Link" is a good choice of wording here. This name reflects some requirements for using this library in compilation process, but I think they are false in general. 
For example, in `no-rdc` mode, there is no "link" step as the code is self-contained by definition.

I think something like "SYCL offload finalization library" better reflects the purpose of this code.

Same applies to `SYCLPostLinkProcess` function from llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp [hunk 2]
```diff
@@ -743,76 +743,76 @@ runSYCLPostLinkTool(ArrayRef<StringRef> InputFiles, const ArgList &Args) {
...
+runSYCLPostLinkLibrary(ArrayRef<StringRef> InputFiles, const ArgList &Args,
+                       module_split::IRSplitMode Mode) {
+  std::vector<module_split::ProcessedModule> OutputModules;
+  llvm::module_split::ModuleProcessingSettings Settings;
   Settings.Mode = Mode;
   Settings.OutputPrefix = "";
 
   for (StringRef InputFile : InputFiles) {
+    if (DryRun)
+      break;
```
Comment (bader) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:772]: I find the previous version significantly more readable than the new one.

File: llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h [hunk 3]
```diff
@@ -289,35 +291,41 @@ void dumpEntryPoints(const Module &M, bool OnlyKernelsAreEntryPoints = false,
...
-Expected<std::vector<SplitModule>> parseSplitModulesFromFile(StringRef File);
+SmallString<64>
+convertProcessingSettingsToString(const ModuleProcessingSettings &S);
 
-/// Splits the given module \p M according to the given \p Settings.
-Expected<std::vector<SplitModule>>
-splitSYCLModule(std::unique_ptr<Module> M, ModuleSplitterSettings Settings);
+/// Parses the output table file from sycl-post-link tool.
+Expected<std::vector<ProcessedModule>>
+parseProcessedModulesFromFile(StringRef File);
```
Comment (sarnex) [inline llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h:322]: i thought we wanted to seperate module splitting out into a seperate operation, possibly so that we could hook it into any upstream splitting api, but it seems like here we are generalizing these APIs, at least in the name (also, the file name still has split it in which is a little confusing)

what's the long term plan/architecture?

thanks

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 2 above)
Comment (maksimsab) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:772]: Thank you for your thoughts.

Could you help me understand which aspects of the original code you find more readable? These 2 blocks (DryRun, Verbose) are almost identical and they take doubled attention.

File: llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h (hunk 3 above)
Comment (maksimsab) [reply llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h:322]: The following processing pipeline comes only from perspectives of RAM usage and compilation time.
Get new split module -> handle spec constants -> handle ESIMD part -> save bc files to disk -> run SPIRVTranslator.

It is known that keeping split modules in the RAM is not really possible. It leads to the question where we should perform the spec const and ESIMD parts? Before saving to disk or after?

We could do the following scheme:
Split modules -> save bc files to disk -> read them again, perform spec const, ESIMD processing -> save bc files to disk again -> run SPIRVTranslator.
This scheme entails twice disk ops compared to the previous one.

The original scheme takes its roots from the usage of SPIRVTranslator. In case of usage of SPIRV backend it would be much easier like the following:
Split modules and save bc files -> read bc file, perform spec const, ESIMD processing -> run SPIRV backend -> save spirv file to disk.
I expect to move to this scheme once SPIRV backend becomes the main path. However, I don't know when it will happen.

I see your concerns that you need some API for thin-LTO. I was considering to come up with a splitting Pass similar to one that AMD have in llvm-project.

While SPIRVTranslator is the main tool you still need to handle spec constants and ESIMD somewhere in thin-LTO. Most likely, `post-split` processing part might be extracted in order to be invoked from the library and thin-LTO framework.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 1 above)
Comment (maksimsab) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:754]: `SYCLOffloadFinalize` ?

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 2 above)
Comment (bader) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:772]: Okay, can we resolve this duplication by moving the duplicated code to a function? 
These two lines technically skip the whole loop, but it looks like we enter it. Now it's much more difficult to understand what happens when `DryRun` is `true`. The previous version has all code for that in a single scope statement. Now, I need to parse all the code of the function, which is written in unusual style.

File: llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h (hunk 3 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h:322]: thanks for the explanation. actually for thin-lto we call sycl-post-link early (in -c phase) and run processing there. that is actually enough for everything besides spec constants (that we know of right now), so that one i will do as part of device link, so i don't think there is a need for an API to call all sycl-post-link post-split processing for thinlto, at least not at the moment

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 1 above)
Comment (maksimsab) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:774]: In fact, in `no-rdc` mode this processing can be done in `-c` phase because they are already implemented as transformation passes. In that case, there is no need to invoke this library at all in `clang-linker-wrapper`.

Comment (maksimsab) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:774]: Friendly ping, @bader

Comment (bader) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:774]: Friendly pong.

> SYCLOffloadFinalize ?

Could you clarify what are you trying to ask here?

Comment (maksimsab) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:774]: Do you think that `SYCLPostLinkProcess` is better to be replaced by `SYCLOffloadFinalize`?

What do you think about my comment regarding `no-rdc`?

Comment (bader) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:774]: > Do you think that `SYCLPostLinkProcess` is better to be replaced by `SYCLOffloadFinalize`?

`SYCLOffloadFinalize` doesn't seem to fit perfectly, but it sounds better than `SYCLPostLinkProcess` in my opinion.

> What do you think about my comment regarding `no-rdc`?

I agree, but I don't get the motivation to have a separate "library". Correct me if I wrong, but all this library does is running LLVM passes. My understanding is that we can integrate these LLVM passes either into standard optimization pipeline (preferred) or by building a separate optimization pipeline. @sarnex is using this kind of approach in https://github.com/intel/llvm/pull/15083/files#diff-f7dc51acede945c7c8481147e773e9c09230a93b937bee9eab08ee52826cfa8eR1779.

Basically, I would consider integrating new passes into LLVM optimization pipeline for both `rdc` and `no-rdc` modes.

Comment (maksimsab) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:774]: Module splitting can't be integrated into `-c` phase even if it were be implemented as a Pass.

Eventually, I see the compilation pipeline as thin-LTO + optimization pipeline. But I don't see that it happens soon. The plan is to move `sycl-post-link` functionality in `clang-linker-wrapper`. Then move to the thin-LTO once it is ready.
